crates/hir-ty/src/mir/lower.rs

```diff
@@ -0,0 +1,1214 @@
+                    }
+                }
+                match tail {
+                    Some(tail) => self.lower_expr_to_place(*tail, place, current),
+                    None => Ok(current),
+                }
+            }
+            Expr::Loop { body, label } => self.lower_loop(prev_block, *label, |this, begin, _| {
+                let (_, block) = this.lower_expr_to_some_place(*body, begin)?;
+                this.set_goto(block, begin);
+                Ok(())
+            }),
+            Expr::While { condition, body, label } => {
+                self.lower_loop(prev_block, *label, |this, begin, end| {
+                    let (cond_place, to_switch) =
+                        this.lower_expr_to_some_place(*condition, begin)?;
+                    let after_cond = this.new_basic_block();
+                    this.set_terminator(
+                        to_switch,
+                        Terminator::SwitchInt {
+                            discr: Operand::Copy(cond_place),
+                            targets: SwitchTargets::static_if(1, after_cond, end),
+                        },
+                    );
+                    let (_, block) = this.lower_expr_to_some_place(*body, after_cond)?;
+                    this.set_goto(block, begin);
+                    Ok(())
+                })
+            }
+            Expr::For { .. } => not_supported!("for loop"),
+            Expr::Call { callee, args, .. } => {
+                let callee_ty = self.expr_ty(*callee);
+                match &callee_ty.data(Interner).kind {
+                    chalk_ir::TyKind::FnDef(..) => {
+                        let func = Operand::from_bytes(vec![], callee_ty.clone());
+                        self.lower_call(func, args.iter().copied(), place, prev_block)
+                    }
+                    TyKind::Scalar(_)
+                    | TyKind::Tuple(_, _)
+                    | TyKind::Array(_, _)
+                    | TyKind::Adt(_, _)
+                    | TyKind::Str
+                    | TyKind::Foreign(_)
+                    | TyKind::Slice(_) => {
+                        return Err(MirLowerError::TypeError("function call on data type"))
+                    }
+                    TyKind::Error => return Err(MirLowerError::MissingFunctionDefinition),
+                    TyKind::AssociatedType(_, _)
+                    | TyKind::Raw(_, _)
+                    | TyKind::Ref(_, _, _)
+                    | TyKind::OpaqueType(_, _)
+                    | TyKind::Never
+                    | TyKind::Closure(_, _)
+                    | TyKind::Generator(_, _)
+                    | TyKind::GeneratorWitness(_, _)
+                    | TyKind::Placeholder(_)
+                    | TyKind::Dyn(_)
+                    | TyKind::Alias(_)
+                    | TyKind::Function(_)
+                    | TyKind::BoundVar(_)
+                    | TyKind::InferenceVar(_, _) => not_supported!("dynamic function call"),
```

Comment:
hopefully bound vars / inference vars shouldn't show up here, and would indicate a bug?

Reply:
Yes, actually I used `not_supported!` for things that should ideally never happen as well. Long term, `not_supported` should be removed either by supporting or by using a proper error.